nilearn/plotting/surf_plotting.py

```diff
@@ -29,10 +31,502 @@
+        istart = int(norm(-threshold, clip=True) * (our_cmap.N - 1))
+        istop = int(norm(threshold, clip=True) * (our_cmap.N - 1))
+        for i in range(istart, istop):
+            cmaplist[i] = (0.5, 0.5, 0.5, 1.)
+    our_cmap = LinearSegmentedColormap.from_list(
+        'Custom cmap', cmaplist, our_cmap.N)
+    x = np.linspace(0, 1, 100)
+    rgb = our_cmap(x, bytes=True)[:, :3]
+    rgb = np.array(rgb, dtype=int)
+    colors = []
+    for i, col in zip(x, rgb):
+        colors.append([np.round(i, 3), "rgb({}, {}, {})".format(*col)])
+    return colors, our_cmap, norm, vmin, vmax
+
+
+def _configure_title_plotly(title=None,
+                            font_family="Courier New, monospace",
+                            font_color="RebeccaPurple",
+                            font_size=22, pos_x=0.5, pos_y=0.96):
+    """Helper function for plot_surf with plotly engine.
+
+    This function configures the title if provided.
+    """
+    if title is None:
+        return dict()
+    if "title" in LAYOUT:
+        raise ValueError("Title is already configured.")
+    return {"text": title,
+            "font": {"family": font_family,
+                     "size": font_size,
+                     "color": font_color,
+                     },
+            "y": pos_y,
+            "x": pos_x,
+            "xanchor": "center",
+            "yanchor": "top"}
+
+
+def _plot_surf_plotly(surf_mesh, surf_map=None, bg_map=None,
+                      hemi='left', view='lateral', cmap=None,
+                      colorbar=False, threshold=None, vmin=None,
+                      vmax=None, title=None, output_file=None):
+    """Helper function for plot_surf.
+
+    .. versionadded:: 0.8.1
+
+    This function handles surface plotting when the selected
+    engine is plotly.
+
+    .. note::
+        This function assumes that plotly and kaleido are
+        installed.
+
+    """
+    try:
+        import plotly.graph_objects as go
+    except ImportError:
+        raise ImportError("Using engine='plotly' requires that "
+                          "plotly and kaleido are installed.")
+    coords, faces = load_surf_mesh(surf_mesh)
+    x, y, z = coords.T
+    i, j, k = faces.T
+    if cmap is None:
+        cmap = cold_hot
+    vertexcolor = None
+    if bg_map is not None:
+        bg_data = load_surf_data(bg_map)
+        if bg_data.shape[0] != coords.shape[0]:
+            raise ValueError('The bg_map does not have the same number '
+                             'of vertices as the mesh.')
+    if surf_map is not None:
+        _check_surf_map(surf_map, coords.shape[0])
+        colors, our_cmap, norm, vmin, vmax = _colorscale_plotly(
+            cmap, surf_map, threshold, vmin, vmax
+        )
+        if not colorbar:
+            vertexcolor = _get_vertexcolor(
+                surf_map, our_cmap, norm, None, bg_map
+            )
+        else:
+            mesh_3d = go.Mesh3d(x=x, y=y, z=z, i=i, j=j, k=k,
```

Comment:
couldn't we use `vertexcolor` in all cases, even when a colorbar is needed?

Reply:
the background map seems to be ignored in this case

Reply:
>couldn't we use vertexcolor in all cases, even when a colorbar is needed?

At this point I am unsuccessful at displaying a colorbar with vertexcolor in `mesh3d`. I might have missed something, but it seems to work only when setting the `intensity` and `colorscale` attributes... This is not a problem in most cases, but when having both a thresholded stat map and a background, we need to combine the colors of both the map and background, which is exactly what the function `_get_vertexcolor` does. I'm not sure how to do the same thing through intensity and colorscale though.
Pinging @alexisthual who might have an idea here

Reply:
> At this point I am unsuccessful at displaying a colorbar with vertexcolor in `mesh3d`. I might have missed something, but it seems to work only when setting the `intensity` and `colorscale` attributes...

when we specify a color for each node rather than a colormap and values plotly does not know how to create a colorbar, but we can easily add it by making an invisible plot using that colormap. This is what is done for the current nilearn plots: https://github.com/nilearn/nilearn/blob/cc99fa76df137e9ac6010c311e15fa9ac12ce66d/nilearn/plotting/data/js/surface-plot-utils.js#L144
so the current approach is to use this trick and vertexcolors in all cases, which allows to handle everything in the same way and have a background map and a colorbar easily.